extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/db/SQLBasicAuthorizerStorageConnector.java

```diff
@@ -0,0 +1,915 @@
+  }
+
+  private class PermissionsMapper implements ResultSetMapper<Map<String, Object>>
+  {
+    @Override
+    public Map<String, Object> map(int index, ResultSet resultSet, StatementContext context)
+        throws SQLException
+    {
+      int id = resultSet.getInt("id");
+      byte[] resourceJson = resultSet.getBytes("resource_json");
+      try {
+        final ResourceAction resourceAction = jsonMapper.readValue(resourceJson, ResourceAction.class);
+        return ImmutableMap.of(
+            "id", id,
+            "resourceAction", resourceAction
+        );
+      }
+      catch (IOException ioe) {
+        throw new RuntimeException(ioe);
+      }
+    }
+  }
+
+  @Override
+  public List<Map<String, Object>> getPermissionsForRole(String dbPrefix, String roleName)
+  {
+    final String permissionTableName = getPrefixedTableName(dbPrefix, PERMISSIONS);
+
+    return getDBI().inTransaction(
+        new TransactionCallback<List<Map<String, Object>>>()
+        {
+          @Override
+          public List<Map<String, Object>> inTransaction(Handle handle, TransactionStatus transactionStatus)
+              throws Exception
+          {
+            int roleCount = getRoleCount(handle, dbPrefix, roleName);
+            if (roleCount == 0) {
+              throw new BasicSecurityDBResourceException("Role [%s] does not exist.", roleName);
+            }
+
+            return handle
+                .createQuery(
+                    StringUtils.format(
+                        "SELECT %1$s.id, %1$s.resource_json\n"
+                        + "FROM %1$s\n"
+                        + "WHERE %1$s.role_name = :roleName",
+                        permissionTableName
+                    )
+                )
+                .map(permMapper)
+                .bind("roleName", roleName)
+                .list();
+          }
+        }
+    );
+  }
+
+  @Override
+  public List<Map<String, Object>> getPermissionsForUser(String dbPrefix, String userName)
+  {
+    final String roleTableName = getPrefixedTableName(dbPrefix, ROLES);
+    final String userRoleTableName = getPrefixedTableName(dbPrefix, USER_ROLES);
+    final String permissionTableName = getPrefixedTableName(dbPrefix, PERMISSIONS);
+
+    return getDBI().inTransaction(
+        new TransactionCallback<List<Map<String, Object>>>()
+        {
+          @Override
+          public List<Map<String, Object>> inTransaction(Handle handle, TransactionStatus transactionStatus)
+              throws Exception
+          {
+            int userCount = getUserCount(handle, dbPrefix, userName);
+            if (userCount == 0) {
+              throw new BasicSecurityDBResourceException("User [%s] does not exist.", userName);
+            }
+
+            return handle
+                .createQuery(
+                    StringUtils.format(
+                        "SELECT %1$s.id, %1$s.resource_json, %2$s.name\n"
+                        + "FROM %1$s\n"
+                        + "JOIN %2$s\n"
```

Comment:
These joins are another reason to consider fewer tables. They can get really out of hand perfwise.